test/util/assertAriaRole.js

```diff
@@ -0,0 +1,36 @@
+'use strict';
+
+const { By } = require('selenium-webdriver');
+const assert = require('assert');
+
+/**
+ * Confirm the aria-labelledby element.
+ *
+ * @param {obj} t             - ava execution object
+ * @param {String} exampleId  - the id of the element containing the example
+ * @param {String} role       - an aria role
+ * @param {String} roleCount  - the number of elements with role set for example
+ * @param {String} elementTag - the element the role should be found on
+ */
+
+module.exports = async function assertAriaRole (t, exampleId, role, roleCount, elementTag) {
+  const elementSelector = '#' + exampleId + ' [role="' + role + '"]';
+
+  const elements = await t.context.session.findElements(By.css(elementSelector));
+
+  assert.equal(
+    elements.length,
+    roleCount,
+    roleCount + 'role="' + role + '" elements should be found by selector: ' + elementSelector
+  );
+
+  for (let element of elements) {
+    assert.equal(
+      await element.getTagName(),
+      elementTag,
+      'role="' + role + '" should be found on "' + elementTag + '" elements only'
```

Comment:
In all previous tests I have been testing whether the role is found on the element listed in the attribute table as well. You can see in this [example](https://w3c.github.io/aria-practices/examples/tabs/tabs-1/tabs.html) -- every entry, including roles, has a corresponding element. In result, the test would only be as misleading to a developer (about the difference between the example's implementation of the pattern and the pattern itself) as the tables are on the example page.

But your concern does hark back to the question of test flexibility vs rigidity. Maybe we should be more flexible here. The examples typically won't break just because element of the role changes, so perhaps it is not an interesting or necessary test for most examples.

Do you have a final recommendation here, @tatermelon? Maybe @mcking65 should weigh in as well. Should I remove tests of the role's element from all previous tests and instead only verify that the role exists on the expected number of elements?

Reply:
Unsure ... simplicity is assert only what is necessary.

One thing we may want to do is change the wording of the message so it is more like "role tab is expected on element button in this implementation." instead of saying that it should only be found on role button, which is not true.

In some cases we may have specific dependencies on html elements in the example. For instance, if the tablist were on a ul, then we would need role presentation on child LI elements. But, that is not necessary if you have a div with role tablist and that div contains buttons. We have some dependencies like that in the menu examples. Also, if you used a div to make tab elements instead of buttons, then the div that is in the page Tab sequence would require tabindex of 0, but if you use a button element, it does not.

What's most important is that the tab elements are descendants of the tablist element. We have structural requirements like that in all the examples. Typically, changing the specific html element used will not change the structural releationships in the DOM.

We want to lean toward simplicity, even if that adds brittleness. At the same time, we do not want to make unnecessary assertions. We also do not want failure messages that imply a requirement that does not exist.